File: app/src/main/java/com/pioneer/aaron/dolly/fork/calllog/ForkVvmActivity.java
package com.pioneer.aaron.dolly.fork.calllog;

import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.util.Log;
import android.view.View;
import android.view.ViewStub;
import android.widget.Button;
import android.widget.EditText;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.Spinner;

import com.pioneer.aaron.dolly.R;
import com.pioneer.aaron.dolly.fork.DataBaseOperator;
import com.pioneer.aaron.dolly.utils.ExpandCollapseAnimation;
import com.pioneer.aaron.dolly.utils.ForkConstants;

import java.util.HashMap;

import me.yokeyword.fragmentation_swipeback.SwipeBackActivity;

/**
 * Created by Aaron on 11/8/17.
 */

public class ForkVvmActivity extends SwipeBackActivity implements IForkCallLogContract.View {
    private static final String TAG = "ForkVvmActivity";

    private IForkCallLogContract.Presenter mPresenter;
    private HashMap<String, Boolean> mColumnsExist;

    private EditText mPhoneNumberEditText;
    private Button mStartForkButton;
    private ViewStub mAdvancedViewStub;
    private View mAdvancedView;
    private boolean mAdvancedOptViewOpened = false;
    private Button mConfigButton;
    private Spinner mConfigSpinner;
    private Button mDataButton;
    private Spinner mDataSpinner;
    private Button mNotificationButton;
    private Spinner mNotificationSpinner;
    private RadioGroup mSubscriptionRadioGroup;
    private RadioButton mSubOneRadioButton;
    private RadioButton mSubTwoRadioButton;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_forkvvm);

        mPresenter = new ForkCallLogPresenter(this, this);
        mPresenter.loadResInBackground(this);
        if (mPresenter.checkPermissions(this)) {
            initUI();
        }
    }

    private void initUI() {
        if (isTaskRoot()) {
            setSwipeBackEnable(false);
        }
        mColumnsExist = mPresenter.getColumnsExist(this);

        mPhoneNumberEditText = (EditText) findViewById(R.id.call_log_number_edtxt);
        mStartForkButton = (Button) findViewById(R.id.start_fork_calllog_btn);
        mStartForkButton.setOnClickListener(mOnClickListener);
        mStartForkButton.setOnLongClickListener(mOnLongClickListener);
        mAdvancedViewStub = (ViewStub) findViewById(R.id.advanced_vvm_viewstub);
        mSubscriptionRadioGroup = (RadioGroup) findViewById(R.id.subscription_id_group);
        mSubOneRadioButton = (RadioButton) findViewById(R.id.sub_one);
        mSubOneRadioButton.setChecked(true);
        mSubTwoRadioButton = (RadioButton) findViewById(R.id.sub_two);
    }

    private View.OnClickListener mOnClickListener = new View.OnClickListener() {
        @Override
        public void onClick(View v) {
            switch (v.getId()) {
                case R.id.start_fork_calllog_btn:
                    mPresenter.forkVvmCallLog(ForkVvmActivity.this, mPhoneNumberEditText.getText().toString(),
                            mSubOneRadioButton.isChecked() ? ForkConstants.SIM_ONE : ForkConstants.SIM_TWO);
                    break;
                case R.id.config_btn:
                    mPresenter.sendVvmState(ForkCallLogPresenter.VVM_STATE.CONFIGURATION, mConfigSpinner.getSelectedItem().toString());
                    break;
                case R.id.data_btn:
                    mPresenter.sendVvmState(ForkCallLogPresenter.VVM_STATE.DATA, mDataSpinner.getSelectedItem().toString());
                    break;
                case R.id.notification_btn:
                    mPresenter.sendVvmState(ForkCallLogPresenter.VVM_STATE.NOTIFICATION, mNotificationSpinner.getSelectedItem().toString());
                    break;
                default:
                    break;
            }
        }
    };

    private View.OnLongClickListener mOnLongClickListener = new View.OnLongClickListener() {
        @Override
        public boolean onLongClick(View v) {
            switch (v.getId()) {
                case R.id.start_fork_calllog_btn:
                    mPresenter.vibrate();
                    loadAdvancedView();
                    animateAdancedOptView(!mAdvancedOptViewOpened);
                    break;
                default:
                    break;
            }
            return true;
        }
    };

    private void loadAdvancedView() {
        if (mAdvancedViewStub == null) {
            Log.d(TAG, "AdvancedViewStub has been inflated.");
            return;
        }

        mAdvancedView = mAdvancedViewStub.inflate();
        mAdvancedViewStub = null;
        // configuration state
        mConfigButton = (Button) mAdvancedView.findViewById(R.id.config_btn);
        mConfigButton.setOnClickListener(mOnClickListener);
        mConfigSpinner = (Spinner) mAdvancedView.findViewById(R.id.states_spinner);

        // data state
        mDataButton = (Button) findViewById(R.id.data_btn);
        mDataButton.setOnClickListener(mOnClickListener);
        mDataSpinner = (Spinner) findViewById(R.id.data_spinner);

        // notification state
        mNotificationButton = (Button) findViewById(R.id.notification_btn);
        mNotificationButton.setOnClickListener(mOnClickListener);
        mNotificationSpinner = (Spinner) findViewById(R.id.notification_spinner);
    }

    /**
     * @param open TRUE to open, while FALSE to close.
     */
    private void animateAdancedOptView(boolean open) {
        if (mAdvancedView != null) {
            mAdvancedOptViewOpened = !mAdvancedOptViewOpened;
            ExpandCollapseAnimation animation = new ExpandCollapseAnimation(mAdvancedView,
                    open ? ExpandCollapseAnimation.EXPANEDED : ExpandCollapseAnimation.COLLAPSED);
            animation.setDuration(200);
            mAdvancedView.startAnimation(animation);
        } else {
            Log.d(TAG, "Failed to animate advanced opt view due to null object.");
        }
    }

    @Override
    public void toast(String msg) {
        Snackbar.make(mStartForkButton, msg, Snackbar.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/pioneer/aaron/dolly/fork/calllog/ForkVvmActivity.kt
package com.pioneer.aaron.dolly.fork.calllog

import android.graphics.Color
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.util.Log
import android.view.View
import android.view.ViewStub
import android.widget.*
import anko.Id
import anko.immersiveToolbar
import anko.subscriptionLayout
import com.pioneer.aaron.dolly.R
import com.pioneer.aaron.dolly.utils.ExpandCollapseAnimation
import com.pioneer.aaron.dolly.utils.ForkConstants
import me.yokeyword.fragmentation_swipeback.SwipeBackActivity
import org.jetbrains.anko.*
import org.jetbrains.anko.appcompat.v7.toolbar

/**
 * Created by Aaron on 11/8/17.
 */

class ForkVvmActivity : SwipeBackActivity(), IForkCallLogContract.View {
    companion object {
        private const val TAG = "ForkVvmActivity"
    }

    private lateinit var mPresenter: IForkCallLogContract.Presenter

    private var mAdvancedOptViewOpened = false
    private lateinit var mPhoneNumberEditText: EditText
    private lateinit var mStartForkButton: Button
    private var mAdvancedViewStub: ViewStub? = null
    private var mAdvancedView: View? = null
    private lateinit var mConfigButton: Button
    private lateinit var mConfigSpinner: Spinner
    private lateinit var mDataButton: Button
    private lateinit var mDataSpinner: Spinner
    private lateinit var mNotificationButton: Button
    private lateinit var mNotificationSpinner: Spinner
    private lateinit var mSubscriptionRadioGroup: RadioGroup
    private lateinit var mSubOneRadioButton: RadioButton
    private lateinit var mSubTwoRadioButton: RadioButton
    private lateinit var mCallLogNumberInputLayout: ViewStub

    private val mOnClickListener = View.OnClickListener { v ->
        when (v.id) {
            R.id.config_btn -> mPresenter.sendVvmState(ForkCallLogPresenter.VVM_STATE.CONFIGURATION, mConfigSpinner.selectedItem.toString())
            R.id.data_btn -> mPresenter.sendVvmState(ForkCallLogPresenter.VVM_STATE.DATA, mDataSpinner.selectedItem.toString())
            R.id.notification_btn -> mPresenter.sendVvmState(ForkCallLogPresenter.VVM_STATE.NOTIFICATION, mNotificationSpinner.selectedItem.toString())
            else -> {
            }
        }
    }

    private val mOnLongClickListener = View.OnLongClickListener { v ->
        when (v) {
            mStartForkButton -> {
                mPresenter.vibrate()
                loadAdvancedView()
                animateAdvancedOptView(!mAdvancedOptViewOpened)
            }
            else -> {
            }
        }
        true
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        anko.makeImmersive(window)
        window.decorView.systemUiVisibility = View.SYSTEM_UI_FLAG_LAYOUT_STABLE or
                View.SYSTEM_UI_FLAG_LAYOUT_FULLSCREEN
        window.statusBarColor = Color.TRANSPARENT
        ForkVvmActivityUI().setContentView(this)

        mPresenter = ForkCallLogPresenter(this, this)
        if (checkPermission(this)) {
            initUI()
        }
    }

    private fun initUI() {
        if (isTaskRoot) {
            setSwipeBackEnable(false)
        }

        mCallLogNumberInputLayout.inflate()

        mPhoneNumberEditText = findViewById<View>(R.id.call_log_number_edtxt) as EditText
        mSubOneRadioButton = findViewById<View>(Id.subOneRadioButton) as RadioButton
        mSubOneRadioButton.isChecked = true
        mSubTwoRadioButton = findViewById<View>(Id.subTwoRadioButton) as RadioButton
    }

    private fun loadAdvancedView() {
        if (mAdvancedViewStub == null) {
            Log.d(TAG, "AdvancedViewStub has been inflated.")
            return
        }

        mAdvancedView = mAdvancedViewStub!!.inflate()
        mAdvancedViewStub = null
        // configuration state
        mConfigButton = mAdvancedView?.findViewById<View>(R.id.config_btn) as Button
        mConfigButton.setOnClickListener(mOnClickListener)
        mConfigSpinner = mAdvancedView?.findViewById<View>(R.id.states_spinner) as Spinner

        // data state
        mDataButton = findViewById<View>(R.id.data_btn) as Button
        mDataButton.setOnClickListener(mOnClickListener)
        mDataSpinner = findViewById<View>(R.id.data_spinner) as Spinner

        // notification state
        mNotificationButton = findViewById<View>(R.id.notification_btn) as Button
        mNotificationButton.setOnClickListener(mOnClickListener)
        mNotificationSpinner = findViewById<View>(R.id.notification_spinner) as Spinner
    }

    /**
     * @param open TRUE to open, while FALSE to close.
     */
    private fun animateAdvancedOptView(open: Boolean) {
        mAdvancedView?.let {
            mAdvancedOptViewOpened = !mAdvancedOptViewOpened
            val animation = ExpandCollapseAnimation(mAdvancedView!!,
                    if (open) ExpandCollapseAnimation.EXPANEDED else ExpandCollapseAnimation.COLLAPSED)
            animation.duration = 200
            mAdvancedView!!.startAnimation(animation)
        } ?: run {
            Log.d(TAG, "Failed to animate advanced opt view due to null object.")
        }
    }

    override fun toast(msg: String) {
        Snackbar.make(mStartForkButton, msg, Snackbar.LENGTH_SHORT).show()
    }

    inner class ForkVvmActivityUI : AnkoComponent<ForkVvmActivity> {
        override fun createView(ui: AnkoContext<ForkVvmActivity>): View =
                with(ui) {
                    verticalLayout {
                        immersiveToolbar(this@ForkVvmActivity)

                        scrollView {
                            verticalLayout {
                                padding = dimen(R.dimen.activity_horizontal_margin)

                                mCallLogNumberInputLayout = viewStub {
                                    layoutResource = R.layout.call_log_number_input_layout
                                }

                                mSubscriptionRadioGroup = subscriptionLayout(context)

                                mAdvancedViewStub = viewStub {
                                    layoutResource = R.layout.advanced_opt_vvm_layout
                                }

                                mStartForkButton = button {
                                    text = getString(R.string.start_fork_call_logs)
                                    allCaps = false
                                    setOnClickListener {
                                        mPresenter.forkVvmCallLog(this@ForkVvmActivity, mPhoneNumberEditText.text.toString(),
                                                if (mSubOneRadioButton.isChecked) ForkConstants.SIM_ONE else ForkConstants.SIM_TWO)
                                    }
                                    setOnLongClickListener(mOnLongClickListener)
                                }
                            }
                        }

                    }
                }

    }
}
